esp8266: Add disable_irq/enable_irq to machine module

dpgeorge opened the pull request:
Here is an attempt at implementing the functions to control irqs.  The idea is that with irqs disabled everything is quiet and timing is predictable.  Hence the need to disable the ets_loop_iter code when irqs are disabled.

Because irqs may already be disabled when you use these functions the correct usage pattern is:

```
    irq_state = machine.disable_irq()
    # do stuff
    machine.enable_irq(irq_state)
```

But, if we were willing to change the machine API, we could actually do it all with only 1 function, ie:

```
    irq_state = machine.irq_active(False)
    # do stuff
    machine.irq_active(irq_state)
```

(And in a more general case you'd have irq priority levels and be able to raise/lower the levels, eg using machine.irq_priority.)

I guess this is all up for discussion, so this is just a starting point to kick such a discussion off.

See #2050.  And note the suggestion there to have a context manager to control irqs (instead of always having to write a try: finally: block).

dpgeorge committed:
esp8266/modmachine: Add disable_irq and enable_irq functions.

dpgeorge committed:
docs/machine: Make disable_irq and enable_irq docs available for all.

pfalcon commented:
> (And in a more general case you'd have irq priority levels and be able to raise/lower the levels, eg using machine.irq_priority.)

Exactly. So, we need to change the docs saying that disable_irq() returns an opaque value representing previous state of interrupt, which can be passed back to enable_irq(). In parentheses we can add that it can be a boolean, or numeric level, or bitmask, or whatever.

> But, if we were willing to change the machine API, we could actually do it all with only 1 function, ie:

This is too brave I'be afraid. Not talking that it requires changing machine API (and there're many changes being applied to it anyway), it also breaks away from a common concept that interrupt disable/enabled, to which a lot of folks are used. So, I'm -0.25 on changing this (my real concern, as I said is that there's a lot of more important things to change).

pfalcon commented:
> ets_intr_lock();

You're aware that this doesn't really disable all interrupts completely, just sets a particular IRQ level (2 IIRC).

dpgeorge commented:
> So, we need to change the docs saying that disable_irq() returns an opaque value representing previous state of interrupt, which can be passed back to enable_irq().

I agree.  This is how stmhal's IRQ functions are defined at the C level (values are opaque).

Then enable_irq() should be forced to take an argument, to force people to use it properly.  Otherwise, it's tricky to define what enable_irq() does with no arguments.

> You're aware that this doesn't really disable all interrupts completely, just sets a particular IRQ level (2 IIRC).

I didn't know.  But my testing showed that timing was very stable using this function.

pfalcon commented:
So, there're 2 points:

1. HW API should allow to work natively with various IRQ enable/disable models of different CPUs. I'm glad you agree with that.
2. Need to decide whether to use native Xtensa model for esp8266 or some logical abstraction (e.g. True/False like your patch does). Given that we need to plug ets_loop_iter enable/disable into it, logical sounds ok to me, at least for starters (otherwise, we immediately need to answer q's like at which native IRQ level we enable/disable ets_loop).

dpgeorge commented:
> Given that we need to plug ets_loop_iter enable/disable into it,

We could do it at the native level, then ets_loop_iter could detect the native IRQ mask/level and exit immediately if it wasn't at the "all IRQs enabled" level.  After all, ets_loop_iter is a task scheduler, and tasks have less priority compared with interrupts.

pfalcon commented:
> We could do it at the native level, then ets_loop_iter could detect the native IRQ mask/level and exit immediately if it wasn't at the "all IRQs enabled" level.

Makes sense too.

dpgeorge committed:
esp8266/xtirq: Add xtirq.h for controlling xtensa irqs.

dpgeorge committed:
esp8266/ets_alt_task: Don't run ets_loop_iter if irqs are disabled.

ets_loop_iter processes pending tasks, and tasks are considered lower
priority than interrupts, so tasks shouldn't be processed if interrupts
are disabled.

dpgeorge commented:
I updated this PR.  It now uses low-level xtensa instructions to configure the irq levels.

@pfalcon FYI: ets_intr_lock() sets PS.INTLEVEL to 2 and ets_intr_unlock() sets it to 0 (enable all).  The disable_irq() function I added sets PS.INTLEVEL to 15 (the maximum) which disables all interrupts.

I also changed ets_loop_iter to return immediately if PS.INTLEVEL is not 0 (ie if interrupts are disabled at any level).

Note that machine.enable_irq() *must* take an argument for esp8266 (the previous return value of machine.disable_irq()).  This is to force people to use the pair of functions correctly.

dpgeorge commented:
@pfalcon on a related note: I noticed that ets_post() calls ets_intr_lock() and ets_iter_unlock().  But ets_post() can be called (via system_os_post()) from within ets_loop_iter() when interrupts are supposed to be disabled (eg dupterm callback has this behaviour of posting within a task).  So I'd think that ets_post() should save the old irq state and restore it before exiting, instead of the lock/unlock pair.

pfalcon commented:
> I updated this PR. It now uses low-level xtensa instructions to configure the irq levels.

Looks good (if it works), thanks.

pfalcon commented:
> @pfalcon on a related note: I noticed that ets_post() calls ets_intr_lock() and ets_iter_unlock().

And as I mentioned, that's because the original vendor ets_post() from vendor does pretty much the same (at a particular SDK version, and unless you mix up things too heavily). Yes, it's worth to re-check how that is done - at some point.

roger- commented:
Thanks for this!

Any thoughts on adding a context manager for disabling interrupts and automatically restoring their previous state?

pfalcon commented:
@roger-: That's covered in https://github.com/micropython/micropython/issues/1016 .

pfalcon commented:
I have a feeling that change to ets_poll_iter introduced a regression. My testcase: try to transfer a big file using webrepl_cli.py. I still experience stalls, possibly due to wifi coverage issues (test via a router). When it stalled for a good while, run "port_diag.main()". It hangs on output netif ifconfig settings. I experienced that at least couple of times, in row. One time, I left it in such hanged state for a while, and then switching back to a terminal window (after 5+ mins), I saw it wasn't a hard lock up, and it actually finished port_diag.main()'s output and REPL was working. I commented out changes to ets_poll_iter, and wasn't able to reproduce the issue from the first try.

Suggestions: either test well (many times), or conservatively untie ets_poll_iter from CPU interrupt level state (introduce separate var to enable/disable ets_poll_iter).
